pkg/checkpoint/process.go

```diff
@@ -9,138 +10,542 @@ import (
-// checkpoints as well as any secrets / configMaps which are no longer necessary.
-func process(localRunningPods, localParentPods, apiParentPods, activeCheckpoints, inactiveCheckpoints map[string]*v1.Pod, checkpointerPod CheckpointerPod) (start, stop, remove []string) {
-	// If this variable is filled, then it means we need to remove the pod-checkpointer's checkpoint.
-	// We treat the pod-checkpointer's checkpoint specially because we want to put it at the end of
-	// the remove queue.
-	var podCheckpointerID string
-
-	// We can only make some GC decisions if we've successfully contacted an apiserver.
-	// When apiParentPods == nil, that means we were not able to get an updated list of pods.
-	removeMap := make(map[string]struct{})
-	if len(apiParentPods) != 0 {
-
-		// Scan for inacive checkpoints we should GC
-		for id := range inactiveCheckpoints {
-			// If the inactive checkpoint still has a parent pod, do nothing.
-			// This means the kubelet thinks it should still be running, which has the same scheduling info that we do --
-			// so we won't make any decisions about its checkpoint.
-			// TODO(aaron): This is a safety check, and may not be necessary -- question is do we trust that the api state we received
-			//              is accurate -- and that we should ignore our local state (or assume it could be inaccurate). For example,
-			//              local kubelet pod state will be innacurate in the case that we can't contact apiserver (kubelet only keeps
-			//              cached responses from api) -- however, we're assuming we've been able to contact api, so this likely is moot.
-			if _, ok := localParentPods[id]; ok {
-				glog.V(4).Infof("API GC: skipping inactive checkpoint %s", id)
-				continue
+// apiState represents information returned from the various api endpoints for a given pod.
+type apiState struct {
+	// apiAvailable is true if the apiserver was reachable.
+	apiAvailable bool
+	// apiParent is true if the api parent pod exists.
+	apiParent bool
+	// localRunning is true if the CRI shim reports that the pod is running locally.
+	localRunning bool
+	// localParent is true if the kubelet parent pod exists.
+	localParent bool
+}
+
+// action represents the action to be taken based on the state of a checkpoint.
+type action int
+
+const (
+	// none is the default action of "do nothing".
+	none = iota
+	// start means that the checkpoint should be started.
+	start
+	// stop means that the checkpoint should be stopped.
+	stop
+	// remove means that the checkpoint should be garbage collected.
+	remove
+)
+
+// String() implements fmt.Stringer.String().
+func (a action) String() string {
+	switch a {
+	case none:
+		return "none"
+	case start:
+		return "start"
+	case stop:
+		return "stop"
+	case remove:
+		return "remove"
+	default:
+		return "[unknown action]"
+	}
+}
+
+// state represents the current state of a checkpoint.
+type state interface {
```

Comment:
Was a little confused by the name at first because we also have `apiState`, either change `apiState` to something else like `apiCondition`, or rename this to `podState`?

Reply:
Done (both)